NOTES/HowStuffWorks.md

```diff
@@ -0,0 +1,118 @@
+
+### page reloading
+Ensure that you are running the app webserver in development mode.
+
+Then run another server using `npm run dev-server`.
+It wil rebuild while watching for file changes, and it wil trigger your page to reload afterwards.
+
+### hot reloading
+Ensure that you are running the app webserver in development mode.
+
+Then run another server using `npm run hot-dev-server`.
+It wil rebuild while watching for file changes, and it wil update the currently shown and affected component(s) while keeping their state.
+Note this is experimental, and in some cases you'll need to refresh manually.
+
+
+## How the routes work.
+In React after opening the app and going to some page, there is no actual HTML loaded from the server. React app just replaces a component that acts as a page, including any child components.
+But you'd like to have the URL reflect this, and allow user to use browser history (back/forward). A router takes care of these things. (package react-router)
+
+In this case, the root of your app is not the Application React component.
+This starts at `/lib/server.js` which wil use `/config/app.jsx` which instantiates the router and ultimately uses `/app/mainRoutes.jsx` to load routes.
+You'll find that all pages are subroutes within the `app` route, which instantiates `/app/Application/index.jsx`, which contains a `RouteHander` component that inserts subroute output.
+
+
+## How the server JSON API works.
+The `/lib/server.js` serves the application, but it also serves the API URLs that the stores talk with.
+It initializes two databases (todolist and todoitem) once, and then continues to listen for GET/POST requests on specific URLs.
+
+
+## How the stores work.
+As in Flux; the Stores affect the React components. (package items-store) And the stores talk to the JSON API. (package superagent)
+
+### Stores setup
+The stores are constructed as such: startpoint is `/config/app.jsx` which wil use `/app/mainStores.js`. This then:
+
+- defines routines to handle JSON API read/writes (package superagent), and 
+- sets up a queue that only allows one REST request at a time, and aggregates subsequent changes. 
+  These are then sent as one request.. :question:
+- and ultimately constructs the two respective `ItemStore` objects (based on `/app/mainStoresDescriptions.js`) to which it assigns these routines, queue, and the initial data. This initial data may have been inserted when `/app/prerender.html`.
+- These are then exported back to `/config/app.jsx`, along with a store named Router. This store just holds a `transition` value which is later read in `/Application/index.jsx` to show "loading...' in UI.
+
+### Stores in use
+
+todo :question: this section needs review and should be extended.
+
+Note on `Application.update()`: normally components should not access the stores except for reading in `getState()`, as demonstrated with the Router store. Everything else should be done with actions.
```

Comment:
> How is it triggered to refresh everything from the server,

`config/app.jsx` invalidates store data when you change the page. When the pages read the data again it is invalid and will be refetched.

> how does it propagate changes when the user edits items, and

The component fires an action (from `app/actions.jsx`) on edit. The action is handled in `app/mainStores.jsx` and writes some items on stores. The stores update their internal cache (which is displayed on the page as optimistic update) and do server requests (through their registered handlers in `app/mainStores.jsx`.

> how do values travel before ending up in some components render() ?

component requests values from store in `getState()` -> `app/mainStores.jsx` read (unavailable or invalidated) data from server -> internal items-store cache for the store -> `getState()` -> components `this.state` -> `render()`

> how does the queue combine multiple requests? I cant imagine subsequent add/remove/edits would result in just one rest call.. do they?

items-store store a single `update` per entry. Any subsequent `updateItem()` call causes both updates to be merged (`mergeUpdates` from items-store). Requests from stores to the server are queued (`queueRequest` in `app/mainStores.jsx`). Here only a single ongoing request is allowed. All writes that happen in the meantime are merged into a single write.